src/runtime/storageNG/tests/store-sequence-test.ts

```diff
@@ -0,0 +1,170 @@
+/**
+ * @license
+ * Copyright (c) 2019 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {assert} from '../../../platform/chai-web.js';
+import {Store, StorageMode, ActiveStore, ProxyMessageType, ProxyMessage} from '../store.js';
+import {SequenceTest, ExpectedResponse, SequenceOutput} from '../../testing/sequence.js';
+import {CRDTCountTypeRecord, CRDTCount, CountOpTypes, CountData} from '../../crdt/crdt-count.js';
+import {DriverFactory, Driver, ReceiveMethod, StorageDriverProvider, Exists} from '../drivers/driver-factory.js';
+
+class MockDriver<Data> extends Driver<Data> {
+  receiver: ReceiveMethod<Data>;
+  async read(key: string) { throw new Error("unimplemented"); }
+  async write(key: string, value: {}) { throw new Error("unimplemented"); }
+  registerReceiver(receiver: ReceiveMethod<Data>) {
+    this.receiver = receiver;
+  }
+  async send(model: Data): Promise<boolean> {
+    return true;
+  }
+}
+
+class MockStorageDriverProvider implements StorageDriverProvider {
+  willSupport(storageKey: string) {
+    return true;
+  }
+  driver<Data>(storageKey: string, exists: Exists): Driver<Data> {
+    return new MockDriver<Data>(storageKey, exists);
+  }
+}
+
+describe('Store Flow', async () => {
+  // Tests a model resync request happening synchronously with model updates from the driver
+  it('services a model request and applies 2 models', async () => {
+    const sequenceTest = new SequenceTest<ActiveStore<CRDTCountTypeRecord>>();
+    sequenceTest.setTestConstructor(() => {
+      DriverFactory.clearRegistrationsForTesting();
+      DriverFactory.register(new MockStorageDriverProvider());
+    
+      const store = new Store('string', Exists.ShouldCreate, null, StorageMode.Direct, CRDTCount);
+      const activeStore = store.activate();
+      return activeStore;
+    });    
+
+    const onProxyMessage = sequenceTest.registerInput('onProxyMessage', 1, 
+      {type: ExpectedResponse.Constant, response: true});
+    const onReceive = sequenceTest.registerInput('onReceive', 1, {type: ExpectedResponse.Void}); 
+
+    const send = sequenceTest.registerOutput('driver.send', 
+      {
+        type: ExpectedResponse.Defer,
+        default: true,
+      }, SequenceOutput.Replace);
+      
+    const idVar = sequenceTest.registerVariable(-1);
+    const isSyncRequest = sequenceTest.registerVariable(false);
+    const model = sequenceTest.registerVariable(() => new CRDTCount(), true);
+  
+    const on = sequenceTest.registerOutput('on',
```

Comment:
I don't understand this - is it monitoring `DirectStore.on(callback)`? I'm guessing not, because the response type and value check in onOutput wouldn't make sense. But there aren't any other methods called 'on' in the code under test (store.ts)...

Reply:
It's installing `callback` as a controlled output (i.e. the test pattern gets to decide what `callback` returns and can act / assert on the values passed to `callback`).

Reply:
https://github.com/PolymerLabs/arcs/blob/master/src/runtime/storageNG/store.ts#L143

Reply:
Right, that's the method I mentioned at first. However, it takes a ProxyMessage and returns a numeric id, whereas this output thing talks about a 'true' response and a ProxyMessage value. How do those line up?

Reply:
It actually takes a `ProxyCallback`, which is a function type that takes a `ProxyMessage` and returns a boolean; that's what the value and response in this output thing are modelling.

The numeric id is like a registration key and that's what ends up in idVar.